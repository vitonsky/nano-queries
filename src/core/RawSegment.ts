import { PrimitiveValue, Value } from '../types';

export class RawSegment implements Value<PrimitiveValue> {
	protected readonly value;
	constructor(value: string | number | null) {
		this.value = value;
	}

	public getValue = () => {
		return this.value;
	};
}
