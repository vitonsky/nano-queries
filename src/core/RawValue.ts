export class RawValue {
	protected readonly value;
	constructor(value: string | number | null) {
		this.value = value;
	}

	public getValue = () => {
		return this.value;
	};
}
