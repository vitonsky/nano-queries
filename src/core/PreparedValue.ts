import { RawValue } from './RawValue';

export class PreparedValue extends RawValue {}
