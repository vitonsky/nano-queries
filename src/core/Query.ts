import { IQuery, QueryParameter, QuerySegment, RawQueryParameter } from '../types';
import { PreparedValue } from './PreparedValue';
import { RawSegment } from './RawSegment';

export const filterOutEmptySegments = (segments: RawQueryParameter[]) =>
	segments.filter((segment) => segment !== undefined) as QueryParameter[];

export class Query implements IQuery {
	protected readonly segments: QuerySegment[] = [];
	constructor(...segments: RawQueryParameter[]) {
		if (segments) {
			this.addSegment(...segments);
		}
	}

	/**
	 * Returns query segments number
	 */
	public size() {
		return this.getSegments().length;
	}

	/**
	 * Returns final query that may be preprocessed
	 * Returned query will be used while compile SQL
	 */
	public getSegments() {
		return this.segments;
	}

	/**
	 * Compile query to SQL string and bindings
	 */
	public toSQL() {
		let sql = '';
		const bindings: Array<string | number | null> = [];
		for (const segment of this.getSegments()) {
			if (segment instanceof Query) {
				const data = segment.toSQL();
				sql += data.sql;
				bindings.push(...data.bindings);
				continue;
			}

			if (segment instanceof PreparedValue) {
				sql += '?';
				bindings.push(segment.getValue());
				continue;
			}

			sql += segment.getValue();
		}

		return { sql, bindings };
	}

	protected addSegment(...segments: RawQueryParameter[]) {
		this.segments.push(
			...filterOutEmptySegments(segments).map((segment) => {
				switch (typeof segment) {
					case 'string':
					case 'number':
						return new RawSegment(segment);

					default:
						return segment === null ? new RawSegment(null) : segment;
				}
			}),
		);
	}
}
