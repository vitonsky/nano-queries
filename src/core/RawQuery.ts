import { IQuery, QueryParameter, QuerySegment, RawQueryParameter } from '../types';
import { PreparedValue } from './PreparedValue';
import { RawValue } from './RawValue';

export const filterOutEmptySegments = (segments: RawQueryParameter[]) =>
	segments.filter((segment) => segment !== undefined) as QueryParameter[];

export class RawQuery implements IQuery {
	protected readonly query: QuerySegment[] = [];
	constructor(...query: RawQueryParameter[]) {
		if (query) {
			this.push(...query);
		}
	}

	/**
	 * Returns query segments number
	 */
	public size() {
		return this.exportQuery().length;
	}

	/**
	 * Returns final query that may be preprocessed
	 * Returned query will be used while compile SQL
	 */
	public exportQuery() {
		return this.query;
	}

	/**
	 * Compile query to SQL string and bindings
	 */
	public toSQL() {
		let sql = '';
		const bindings: Array<string | number | null> = [];
		for (const segment of this.exportQuery()) {
			if (segment instanceof RawQuery) {
				const data = segment.toSQL();
				sql += data.sql;
				bindings.push(...data.bindings);
				continue;
			}

			if (segment instanceof PreparedValue) {
				sql += '?';
				bindings.push(segment.getValue());
				continue;
			}

			sql += segment.getValue();
		}

		return { sql, bindings };
	}

	protected push(...queries: RawQueryParameter[]) {
		this.query.push(
			...filterOutEmptySegments(queries).map((segment) => {
				switch (typeof segment) {
					case 'string':
					case 'number':
						return new RawValue(segment);

					default:
						return segment === null ? new RawValue(null) : segment;
				}
			}),
		);
	}
}
