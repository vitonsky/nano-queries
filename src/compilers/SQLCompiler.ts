import { PreparedValue } from '../core/PreparedValue';
import { Query } from '../core/Query';
import { QueryBindings } from '../types';

export interface CommandWithBindings<T> {
	command: string;
	bindings: T[];
}

export interface Compiler<T> {
	compile: (query: Query) => T;
}

export type SQLCompilerConfig = {
	getPlaceholder: (valueIndex: number) => string;
	onPostProcess?: (code: string) => string;
};

export class SQLCompiler implements Compiler<CommandWithBindings<QueryBindings>> {
	private readonly config: SQLCompilerConfig;
	constructor(options?: Partial<SQLCompilerConfig>) {
		this.config = {
			...options,
			getPlaceholder: options?.getPlaceholder ?? (() => '?'),
		};
	}
	/**
	 * Compile query to SQL string and bindings
	 */
	public compile(query: Query): CommandWithBindings<QueryBindings> {
		const sharedState = {
			valueIndex: 0,
		};

		const processQuery = (query: Query): CommandWithBindings<QueryBindings> => {
			let command = '';
			const bindings: Array<string | number | null> = [];
			for (const segment of query.getSegments()) {
				if (segment instanceof Query) {
					const data = processQuery(segment);
					command += data.command;
					bindings.push(...data.bindings);
					continue;
				}

				if (segment instanceof PreparedValue) {
					const placeholder = this.config.getPlaceholder(
						sharedState.valueIndex,
					);
					sharedState.valueIndex++;

					command += placeholder;
					bindings.push(segment.getValue());
					continue;
				}

				command += segment.getValue();
			}

			return { command, bindings };
		};

		const { command, bindings } = processQuery(query);
		return {
			command: this.config.onPostProcess
				? this.config.onPostProcess(command)
				: command,
			bindings,
		};
	}

	/**
	 * Compile query to SQL string and bindings
	 */
	public toSQL = (query: Query): { sql: string; bindings: QueryBindings[] } => {
		const { command: sql, bindings } = this.compile(query);
		return { sql, bindings };
	};
}
