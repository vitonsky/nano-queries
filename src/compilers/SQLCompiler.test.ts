import { format } from 'sql-formatter';

import { PreparedValue } from '../core/PreparedValue';
import { Query } from '../core/Query';
import { RawSegment } from '../core/RawSegment';
import { SQLCompiler } from './SQLCompiler';

test('Compiler can process linear queries', () => {
	const compiler = new SQLCompiler();

	expect(
		compiler.compile(
			new Query(
				new RawSegment('SELECT *'),
				new RawSegment(' '),
				new RawSegment('FROM foo'),
				new RawSegment(' '),
				new RawSegment('WHERE x='),
				new PreparedValue(1),
				new RawSegment(' '),
				new RawSegment('LIMIT'),
				new RawSegment(' '),
				new PreparedValue(2),
				new RawSegment(' '),
				new RawSegment('OFFSET'),
				new RawSegment(' '),
				new PreparedValue(3),
			),
		),
	).toEqual({
		command: 'SELECT * FROM foo WHERE x=? LIMIT ? OFFSET ?',
		bindings: [1, 2, 3],
	});
});

test('Compiler can process nested queries', () => {
	const compiler = new SQLCompiler();

	expect(
		compiler.compile(
			new Query(
				new RawSegment('SELECT *'),
				new RawSegment(' '),
				new RawSegment('FROM foo'),
				new RawSegment(' '),
				new RawSegment('WHERE x='),
				new PreparedValue(1),
				new RawSegment(' AND '),
				new Query(
					new RawSegment('('),
					new RawSegment('SELECT y FROM bar WHERE n='),
					new PreparedValue('foo'),
					new RawSegment(' OR '),
					new Query(new RawSegment('n2='), new PreparedValue('bar')),
					new RawSegment(')'),
				),
				new RawSegment(' '),
				new RawSegment('LIMIT'),
				new RawSegment(' '),
				new PreparedValue(2),
				new RawSegment(' '),
				new RawSegment('OFFSET'),
				new RawSegment(' '),
				new PreparedValue(3),
			),
		),
	).toEqual({
		command:
			'SELECT * FROM foo WHERE x=? AND (SELECT y FROM bar WHERE n=? OR n2=?) LIMIT ? OFFSET ?',
		bindings: [1, 'foo', 'bar', 2, 3],
	});
});

describe('Compiler options', () => {
	test('Compilation with positional placeholders', () => {
		const compiler = new SQLCompiler({
			getPlaceholder(index) {
				return '$' + (index + 1);
			},
		});

		expect(
			compiler.compile(
				new Query(
					new RawSegment('SELECT *'),
					new RawSegment(' '),
					new RawSegment('FROM foo'),
					new RawSegment(' '),
					new RawSegment('WHERE x='),
					new PreparedValue(1),
					new RawSegment(' AND '),
					new Query(
						new RawSegment('('),
						new RawSegment('SELECT y FROM bar WHERE n='),
						new PreparedValue('foo'),
						new RawSegment(' OR '),
						new Query(new RawSegment('n2='), new PreparedValue('bar')),
						new RawSegment(')'),
					),
					new RawSegment(' '),
					new RawSegment('LIMIT'),
					new RawSegment(' '),
					new PreparedValue(2),
					new RawSegment(' '),
					new RawSegment('OFFSET'),
					new RawSegment(' '),
					new PreparedValue(3),
				),
			),
		).toEqual({
			command:
				'SELECT * FROM foo WHERE x=$1 AND (SELECT y FROM bar WHERE n=$2 OR n2=$3) LIMIT $4 OFFSET $5',
			bindings: [1, 'foo', 'bar', 2, 3],
		});
	});

	test('Compilation with post processing', () => {
		const compiler = new SQLCompiler({
			getPlaceholder(index) {
				return '$' + (index + 1);
			},
			onPostProcess(code) {
				return format(code, { language: 'postgresql' });
			},
		});

		expect(
			compiler.compile(
				new Query(
					new RawSegment('SELECT *'),
					new RawSegment(' '),
					new RawSegment('FROM foo'),
					new RawSegment(' '),
					new RawSegment('WHERE x='),
					new PreparedValue(1),
					new RawSegment(' AND '),
					new Query(
						new RawSegment('('),
						new RawSegment('SELECT y FROM bar WHERE n='),
						new PreparedValue('foo'),
						new RawSegment(' OR '),
						new Query(new RawSegment('n2='), new PreparedValue('bar')),
						new RawSegment(')'),
					),
					new RawSegment(' '),
					new RawSegment('ORDER BY embedding <=> [0.1, 0.2, 0.3] ASC'),
					new RawSegment(' '),
					new RawSegment('LIMIT'),
					new RawSegment(' '),
					new PreparedValue(2),
					new RawSegment(' '),
					new RawSegment('OFFSET'),
					new RawSegment(' '),
					new PreparedValue(3),
				),
			),
		).toMatchSnapshot();
	});
});
