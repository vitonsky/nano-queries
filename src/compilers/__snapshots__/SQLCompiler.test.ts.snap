// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Compiler options > Compilation with post processing 1`] = `
{
  "bindings": [
    1,
    "foo",
    "bar",
    2,
    3,
  ],
  "command": "SELECT
  *
FROM
  foo
WHERE
  x = $1
  AND (
    SELECT
      y
    FROM
      bar
    WHERE
      n = $2
      OR n2 = $3
  )
ORDER BY
  embedding <=> [0.1, 0.2, 0.3] ASC
LIMIT
  $4
OFFSET
  $5",
}
`;
